### PR TITLE
perf: skip re-hashing unchanged GGUF sources on model load

### LINKED ISSUE #949: perf: Sidecar hash cache for GGUF model load: eliminate SHA-256 re-hashing on every startup
## Problem

Every GGUF model load recomputes a full-file SHA-256 over all shards, even when the file has not changed since the previous load. This adds **3–40+ seconds** of pure sequential-read wall time (depending on model size and storage) **after every** `synthetic_direct_gguf_package()` call — which fires on every server start, model switch, and any runtime path that rebuilds the `SkippyPackageIdentity`.

### Where it happens

Three independent `file_sha256()` call sites, but the hot path is in the host-runtime packaging layer:

- **`crates/mesh-llm-host-runtime/src/inference/skippy/package.rs:57`** – `synthetic_direct_gguf_package()` is the primary choke point. It calls `source_file()` → `file_sha256()` **per shard**, then `aggregate_source_sha256()` to combine them. Called from `translation.rs:38`, `translation.rs:104`, and `skippy/mod.rs:961`.
- **`crates/skippy-runtime/src/package.rs:1179`** – a separate `file_sha256()` used during layer-package verification.
- **`crates/skippy-model-package/src/main.rs:1896`** – used during package creation (build-time only, not a load-time concern).

### How SHA-256 is used

The two hash values produced are:

1. **`source_model_sha256`** – per-shard SHA-256 combined into one aggregate hash.
2. **`manifest_sha256`** – SHA-256 of a JSON manifest that includes `source_model_sha256` as a field.

These values serve **purely as Rust-side coordination identity keys** — they drive stage deduplication, cache keys, and split topology planning. They never reach the native FFI layer, gossip, or llama.cpp. **No cryptographic security property is enforced.**

### Current read strategy

`file_sha256()` reads each shard sequentially in **64 KiB chunks** via `BufReader<File>` and feeds `sha2::Sha256`:

```rust
fn file_sha256(path: &Path) -> Result<String> {
    let mut reader = BufReader::new(File::open(path)?);
    let mut hasher = Sha256::new();
    let mut buffer = [0u8; 64 * 1024];
    loop {
        let read = reader.read(&mut buffer)?;
        if read == 0 { break; }
        hasher.update(&buffer[..read]);
    }
    Ok(hex_lower(&hasher.finalize()))
}
```

The bottleneck is **disk read throughput**, not CPU hashing speed. Swapping to xxHash or another fast hash does not help — both are I/O bound on a full file scan.

## Proposal: Sidecar Hash Cache

Compute the SHA-256 once per file (on first encounter) and persist it. On subsequent loads, verify that the file size and mtime are unchanged, then reuse the cached hash — skipping the full-file read entirely.

### Design sketch

**Cache record** — stored as a file in `~/.mesh-llm/cache/hashes/` (or alongside the model directory):

```json
{
  "version": 1,
  "entries": [
    {
      "path": "/path/to/model-00001-of-00005.gguf",
      "size": 8589934592,
      "mtime_nanos": 1718580093000000000,
      "sha256": "abc123..."
    }
  ]
}
```

**Lookup flow:**
1. Stat the file → get `(size, mtime_nanos)`.
2. Look up in sidecar cache → if `(size, mtime)` match, return cached SHA-256.
3. No match → compute SHA-256, persist new entry, return.

**Cache scope:** Per-file (shard-level), not per-model. The `aggregate_source_sha256()` combines per-file hashes deterministically, so it benefits automatically.

**False-positive risk:** Matching `(size, mtime)` is sufficient for an identity key — we do not need to cryptographically prove contents. The cache is advisory (performance optimization), not a security boundary.

## Projected Time Savings

Numbers assume a modern NVMe SSD (~4 GB/s sequential read) and typical SATA SSD (~500 MB/s). Savings apply to **every model load after the first encounter**.

| Model size | Current cost (NVMe) | Current cost (SATA) | With cache |
|-----------|-------------------|---------------------|------------|
| 20 GB     | ~5 s              | ~40 s               | ~0 s       |
| 40 GB     | ~10 s             | ~80 s               | ~0 s       |
| 80 GB     | ~20 s             | ~160 s              | ~0 s       |

(Multi-shard models incur additional per-shard stat + read overhead during `direct_gguf_source_files()`, but that is already I/O minimal compared to the full SHA-256 pass.)

## Implementation footprint

- New module: `crates/mesh-llm-host-runtime/src/inference/skippy/hash_cache.rs` (or similar) — ~150 lines.
- Change `source_file()` in `package.rs` to consult the cache before falling back to `file_sha256()`.
- Cache file location via `~/.mesh-llm/cache/hashes/` — respects existing `MESH_LLM_RUNTIME_ROOT` / `XDG_RUNTIME_DIR` conventions.
- No new crate dependencies needed (serde + serde_json already in tree).

## Prior discussion

- Issue #717 (comment) — referenced in `translation.rs:50-58`, noted the double-hash bug that was previously fixed.
- Previous analysis scored this as the **#2 load-time choke point** (score 72/100), behind only the native `skippy_model_open` C++ block (fixed cost, score 100/100).

### PR DESCRIPTION
**Pull request by Abhisheklearn12**
## Title
Skip re-hashing unchanged GGUF models on startup and model switch

## Original problem
Every direct GGUF model load recomputed a full-file SHA-256 over all shards, even when the files had not changed since the last load. This added seconds to minutes of sequential reads to every server start, model switch, and any path that rebuilds the `SkippyPackageIdentity`.

Fixes #949.

## Diagnostics

- Traced hash consumers before changing anything: stage workers echo `manifest_sha256` from the request, and per-file hashes never leave `package.rs`, so the cache has no cross-node or mixed-version impact.
- Measured cache hit cost with a temporary probe: 80 to 125 us on a 1 GiB file. The hit path is one stat plus a small JSON read, so cost does not grow with file size.

## Fix
A sidecar hash cache: one small JSON record per source file (schema version, algo, path, size, mtime_nanos, and the sha256 digest) stored under `~/.mesh-llm/cache/hashes/`. `source_file()` stats the file and reuses the stored hash when size and mtime still match, skipping the full-file read entirely. On a miss it computes once and persists.

Properties:

- Advisory only: any unreadable, corrupt, or mismatched record (schema version, algo, path, size, mtime) falls back to recomputing, and a cache write failure can never fail a model load.
- Writes go through temp file plus rename, so concurrent loads never observe torn records.
- `MESH_LLM_HASH_CACHE_DIR` overrides the location (mirroring `MESH_LLM_RUNTIME_ROOT`). With no home directory and no override, caching is disabled and behavior is identical to today.
- Identity values are byte-for-byte identical to main. No new dependencies, roughly 45% tests.

Measured impact on repeat loads:

| Path | Cost |
|---|---|
| Full SHA-256 read (before, every load) | ~650 ms per GiB (~13 s for a 20 GB model, more on slow disks) |
| Sidecar cache hit (after) | ~100 us per file, independent of file size |

## Why not xxHash

#949 discussed swapping SHA-256 for xxHash. I built it fully (twox-hash 2.1.2) and measured 1 GiB in-memory hashing:

| Build | sha256 (current) | xxh3 |
|---|---|---|
| release | 561 ms | 23 ms (~24x faster) |
| debug | 650 ms | 5.2 s (~8x slower) |

Release confirms the expected win, but debug regresses badly: twox-hash's hot path is `#[inline]` plus generics, so its code monomorphizes into the calling crate and compiles at opt-level 0. The `[profile.dev.package]` opt-level 3 override (the "do not remove this section" block in the root Cargo.toml that keeps sha2 fast in dev builds) cannot reach it. I verified via `cargo test --no-run -v` that the override was actually applied to twox-hash before concluding this.

Shipping it would reintroduce the #717 debug-startup slowness, and since the cache makes repeat loads free, xxHash only helped first encounters. Dropped. A follow-up could put the hash loop in a small optimized leaf crate if the first-encounter win is still wanted.

## Validation

- [x] `cargo check -p mesh-llm-host-runtime` and `cargo check -p mesh-llm`
- [x] `cargo clippy --all-targets -- -D warnings` on both crates
- [x] `cargo test -p mesh-llm-host-runtime --lib`: 1575 passed, 0 failed (12 new tests: cache hit, miss, size and mtime invalidation, corrupt entry, schema and algo mismatch, unwritable dir, overwrite, hash stability)
- [x] `cargo fmt --all --check`
- [x] No UI changes, so no screenshots needed.

## Summary by CodeRabbit

* **Performance Improvements**
  * Reduced repeated SHA-256 hashing for direct GGUF source files by reusing digests from a persisted local hash side-cache.
  * Cached digests are reused only when they exactly match the file’s current metadata and pass strict validation.

* **Reliability**
  * Improved cache correctness by ignoring malformed, corrupt, or stale entries and recomputing when metadata changes.
  * Made cache writes best-effort and safe against partial updates.

* **Tests**
  * Expanded test coverage for cache hits/misses, digest/metadata validation, corruption handling, and timestamp edge cases.

### CODE REVIEW

File: crates/mesh-llm-host-runtime/src/inference/skippy/package.rs [hunk 1]
```diff
@@ -211,14 +221,42 @@ fn source_file(path: &Path) -> Result<SkippyPackageSourceFile> {
...
+    path: &Path,
+    metadata: &std::fs::Metadata,
+    digest_cache: Option<&SidecarDigestCache>,
+) -> Result<String> {
+    let mtime_nanos = hash_cache::file_mtime_nanos(metadata);
+    if let (Some(cache), Some(mtime_nanos)) = (digest_cache, mtime_nanos)
+        && let Some(sha256) = cache.lookup(path, metadata.len(), mtime_nanos)
+    {
+        tracing::debug!(path = %path.display(), "GGUF source sha256 cache hit");
+        return Ok(sha256);
```
[the remark this answers is not in the document]
Comment (Abhisheklearn12) [reply crates/mesh-llm-host-runtime/src/inference/skippy/package.rs:245]: fixed in b4e44e0f. put the validation in `SidecarDigestCache::lookup` like your proposed fix suggested `is_sha256_hex` requires exactly 64 lowercase hex chars, so this call site is covered too since every hit goes through `lookup`. the `cached-sha256` fixture is now a proper 64-char value, and there's a `lookup_misses_on_malformed_digest` test covering short, empty, 63-char and uppercase digests

**Comment by ndizazzo** [comment]
TY! Will be able to review when I'm back home on Wednesday

**Comment by Abhisheklearn12** [comment]
> TY! Will be able to review when I'm back home on Wednesday

Thanks, appreciate it!

**Comment by ndizazzo** [comment]
@Abhisheklearn12 One thing I realized as I was reading through the PR description was that the SHA256 hashing explicitly had it's debug config optimized using the cargo file directive to use full compiler optimizations... I am wondering if the tests for xxHash were not directly fair because of that fact? 

Perhaps it might be worth a quick retest using full compiler optimizations for xxHash as well, or removing the debug optimizations for normal SHA256?

**Comment by Abhisheklearn12** [comment]
> @Abhisheklearn12 One thing I realized as I was reading through the PR description was that the SHA256 hashing explicitly had it's debug config optimized using the cargo file directive to use full compiler optimizations... I am wondering if the tests for xxHash were not directly fair because of that fact?
> 
> Perhaps it might be worth a quick retest using full compiler optimizations for xxHash as well, or removing the debug optimizations for normal SHA256?

@ndizazzo, good instinct to check, i reran the full matrix to be sure. same machine, 1 GiB in-memory, twox-hash 2.1.2:

| config | xxh3 | sha256 |
|---|---|---|
| dev, per-package `opt-level=3` on **both** hash crates (mesh-llm's approach, applied to both) | 5.51 s | 0.60 s |
| dev, no optimizations for either | 5.81 s | 15.32 s |
| dev, but the **calling** crate at opt-level 3 | 0.09 s | 0.58 s |
| release | 0.023 s | 0.57 s |

so the original test was already the fair version: twox-hash had the identical `[profile.dev.package]` override, and i verified via `cargo test --no-run -v` that it really compiled at opt-level 3. the interesting part is rows 1 vs 2. the override rescues sha2 25x (15.3s -> 0.6s) but does nothing for twox-hash (5.8s -> 5.5s). that's because sha2's compress core is non-generic and codegens inside the sha2 crate where the override applies, while twox-hash's hot path is `#[inline]` plus generics, so it monomorphizes into the calling crate and compiles at the caller's opt-level 0.

row 3 is the smoking gun: the hash crates are still fully optimized, but now the calling crate is too, and xxh3 jumps 60x to 90ms. the algorithm was never the problem, it's where the code gets compiled.

on your second idea (dropping sha2's debug override): row 2 shows that makes debug loads dramatically worse for both (sha2 falls to 15.3s/GiB), so the current setup is actually the best debug config available, and it only works because sha2's architecture lets the override land.

net: in the repo's real dev profile, xxh3 loses ~9x in debug and no per-package trick can fix it. the clean path to the release-side 24x win would be a tiny leaf crate owning the hash loop (so codegen lands where an override can reach). happy to do that as a follow-up if you want it.

**Review by ndizazzo — CHANGES_REQUESTED** [review]
Thanks again @Abhisheklearn12! This will be a nice performance improvement overall...

One very minor concern before merging: using only path, size, and mtime can reuse a stale SHA if a GGUF is replaced with different same-size content while its mtime is preserved.

Could we also store and check the file’s ctime? Replacing or modifying a file updates ctime, even if tooling restores its mtime, so this should catch the normal stale-cache case without re-hashing unchanged models.

**Comment by Abhisheklearn12** [comment]
> Thanks again @Abhisheklearn12! This will be a nice performance improvement overall...
> 
> One very minor concern before merging: using only path, size, and mtime can reuse a stale SHA if a GGUF is replaced with different same-size content while its mtime is preserved.
> 
> Could we also store and check the file’s ctime? Replacing or modifying a file updates ctime, even if tooling restores its mtime, so this should catch the normal stale-cache case without re-hashing unchanged models.

@ndizazzo , thank you! done. added ctime to the record and the lookup check, schema bumped to v2 so old entries miss cleanly. also windows std doesn't expose change-time, so it stays size+mtime there. also added a regression test replaying your exact scenario: same-size replace with the mtime restored via `set_modified`, asserting the cache misses because ctime moved

**Review by ndizazzo — APPROVED** [review]

**Comment by michaelneale** [comment]
nice one thanks @Abhisheklearn12
